### DISCUSSION TITLE
Support Concurrency for Steps for more granular locking between workflow runs

### DISCUSSION THREAD

**Discussion by HariSekhon** [community/community#18065] [Actions]
Use Case:

Support `concurrency` for steps, not just workflow and job level, because this would allow for more granular locking between workflow runs, maximizing concurrency between steps that don't affect each other.

Jenkins for example supports locking right down to the individual step level.

This is important for workflows that do lots of different steps such as setup steps, downloading binaries etc. which can be done in parallel and shouldn't be concurrency locked whereas other steps like running Terraform plan/apply do need the locking but only for that exact step, eg:

https://github.com/HariSekhon/GitHub-Actions/blob/master/.github/workflows/terraform.yaml

**Comment by chrispat** [comment]
This isn't possible with actions.  Each job allocates a VM and the steps are run within the workspace on that VM.  Allowing concurrency at the step level would require keeping VMs running for potentially hours which would be a waste of resources and come at potentially a very high cost.

**Reply by HariSekhon** [reply]
That's a good point, long-running locks held by other workflows could require waiting on a compute VM... I get the trade-off.

It's just something that I'm used to using in Jenkins, and in my experience you don't end up waiting hours for another workflow's step to finish, it's usually a wait of minutes for something like a deployment lock. I've used this quite a lot in production to great effect and improved safety by having docker image builds running concurrently in different Jenkins pipelines, and only the deploy stage in each pipeline locks for serialization to minimize the amount of blocking time.

I suspect something similar could be achieved using different workflows for build and deploy and only having the deploy workflow use concurrency locking.

**Comment by HariSekhon** [comment]
Related to:

https://github.com/github-community/community/discussions/18670
